WebDriver: Implement enough `window.open()` to have /window/new endpoint

kalenikaliaksandr opened the pull request:
1) Separate BrowsingContext into "local" (running in current WebContent process) and "remote" (running in another WebContent process).
2) Introduce IPC call to open a new browser tab: required to create new top level browsing. Allows `window.open()` to spawn new tabs.
4) Implement POST `/window/new` WebDriver endpoint.

linusg commented on Userland/Libraries/LibWeb/Crypto/Crypto.cpp:
```diff
@@ -72,6 +72,17 @@ WebIDL::ExceptionOr<String> Crypto::random_uuid() const
+    visitor.visit(m_subtle.ptr());
+}
+
+ErrorOr<String> generate_random_uuid()
```
`// https://w3c.github.io/webcrypto/#dfn-generate-a-random-uuid`

linusg commented on Ladybird/BrowserWindow.cpp:
```diff
@@ -333,7 +333,7 @@ void BrowserWindow::debug_request(DeprecatedString const& request, DeprecatedStr
 }
 
-void BrowserWindow::new_tab(QString const& url, Activate activate)
+Tab* BrowserWindow::new_tab(QString const& url, Activate activate)
```
This is never null and should be a reference

kalenikaliaksandr replied:
yes, fixed.

linusg commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -522,6 +522,44 @@ Messages::WebDriverClient::GetTitleResponse WebDriverConnection::get_title()
+    // FIXME: 4. Let type hint be the result of getting the property "type" from the parameters argument.
+
+    // 5. Create a new top-level browsing context by running the window open steps with url set to "about:blank",
+    // target set to the empty string, and features set to "noopener" and the user agent configured to create a new
+    // browsing context. This must be done without invoking the focusing steps for the created browsing context. If
+    // type hint has the value "tab", and the implementation supports multiple browsing context in the same OS
+    // window, the new browsing context should share an OS window with the current browsing context. If type hint
+    // is "window", and the implementation supports multiple browsing contexts in separate OS windows, the
+    // created browsing context should be in a new OS window. In all other cases the details of how the browsing
+    // context is presented to the user are implementation defined.
```
Please indent to the same level as the `// 5.`

linusg commented on Userland/Libraries/LibWeb/HTML/BrowsingContext.h:
```diff
@@ -302,6 +304,8 @@ class BrowsingContext final
     CSSPixelSize m_size;
     CSSPixelPoint m_viewport_scroll_offset;
 
+    String m_webdriver_handle;
```
`// https://w3c.github.io/webdriver/#dfn-window-handles`

linusg commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -522,6 +527,44 @@ Messages::WebDriverClient::GetTitleResponse WebDriverConnection::get_title()
+    // FIXME: 4. Let type hint be the result of getting the property "type" from the parameters argument.
+
+    // 5. Create a new top-level browsing context by running the window open steps with url set to "about:blank",
+    // 5. target set to the empty string, and features set to "noopener" and the user agent configured to create a new
+    // 5. browsing context. This must be done without invoking the focusing steps for the created browsing context. If
+    // 5. type hint has the value "tab", and the implementation supports multiple browsing context in the same OS
+    // 5. window, the new browsing context should share an OS window with the current browsing context. If type hint
+    // 5. is "window", and the implementation supports multiple browsing contexts in separate OS windows, the
+    // 5. created browsing context should be in a new OS window. In all other cases the details of how the browsing
+    // 5. context is presented to the user are implementation defined.
```
I meant like this:

```
// 5. xxx
//    xxx
//    xxx
...
```

kalenikaliaksandr replied:
I misunderstood comment in a goofiest way, sorry 😅 Fixed!

nico commented:
Should this ref the webdriver tracking bug?

kalenikaliaksandr commented:
yeah, let's reference web driver endpoints list issue https://github.com/SerenityOS/serenity/issues/15551

linusg reviewed (commented):
Looks solid but I'll leave the final review to @trflynn89 :+1:

trflynn89 commented:
I got a crash when I tried this out, hitting this order of endpoints:
```python
driver.post('/session/{session_id}/url', {'url': 'https://xkcd.com'})
driver.get('/session/{session_id}/window/handles')

driver.post('/session/{session_id}/window/new')

driver.post('/session/{session_id}/url', {'url': 'https://serenityos.org'})
driver.get('/session/{session_id}/window/handles')
```

We crash on `/window/new` in `WebDriverConnection::new_window()` here:
```c++
    // 6. Let handle be the associated window handle of the newly created window.
    auto handle = browsing_context->webdriver_handle();
```

trflynn89 commented on Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:
```diff
@@ -83,9 +84,11 @@ JS::NonnullGCPtr<BrowsingContext> BrowsingContext::create_a_new_top_level_browsi
 // https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context
 JS::NonnullGCPtr<BrowsingContext> BrowsingContext::create_a_new_browsing_context(Page& page, JS::GCPtr<DOM::Document> creator, JS::GCPtr<DOM::Element> embedder, BrowsingContextGroup&)
 {
+    String webdriver_handle = Crypto::generate_random_uuid().release_value_but_fixme_should_propagate_errors();
```
Feels a bit wasteful that every browsing context will go through the work of creating a UUID that is unused during normal usage. Maybe WebDriverConnection should instead set this on the Page/BrowsingContext when `set_is_webdriver_active` is called?

kalenikaliaksandr replied:
hmm, I think that makes sense but I expect generating UUID to be pretty cheap operation. Also constructing BC is required only for a new tab/iframes which should not be that frequent.

trflynn89 commented on Userland/Libraries/LibWebView/ViewImplementation.cpp:
```diff
@@ -181,6 +181,8 @@ ErrorOr<NonnullRefPtr<WebView::WebContentClient>> ViewImplementation::launch_web
     auto new_client = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) WebView::WebContentClient(move(socket), *this)));
     new_client->set_fd_passing_socket(TRY(Core::LocalSocket::adopt_fd(ui_fd_passing_fd)));
 
+    m_handle = new_client->get_window_handle();
```
This only applies to Ladybird - can we move this somewhere that Browser also makes use of this?
For example, we could make this a lazy `Optional<String>` class member, where the getter makes the IPC call if it isn't set:
```c++
String const& ViewImplementation::handle() const {
    if (!m_handle.has_value())
        m_handle = client().get_window_handle();
    return *m_handle;
}

mutable Optional<String> m_handle;
````

trflynn89 replied:
Look at the `#if` surrounding this method - this code doesn't exist on Serenity, therefore Browser won't be able to use this.

For reference, this function is for launching the WebContent process from Ladybird. This function doesn't exist on Serenity because launching WebContent happens from a centralized service called SystemServer.

trflynn89 commented on Userland/Libraries/LibWeb/HTML/AbstractBrowsingContext.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace Web::HTML {
+
+class AbstractBrowsingContext : public JS::Cell {
```
Hmm I don't *love* the name `AbstractBrowsingContext`. I think it would be better to have:
`BrowsingContext` -> virtual base
`LocalBrowsingContext` - inherit from `BrowsingContext` for this WC process (what is currently `BrowsingContext`)
`RemoteBrowsingContext` - inherit from `BrowsingContext` for remote WC proccesses

...but that might be a huge pain to rename the existing `BrowsingContext` to `LocalBrowsingContext`, so I won't push too hard here. Maybe I'm just having Java flashbacks :smile:

kalenikaliaksandr replied:
yes, I totally agree. maybe in another PR? :) I would prefer to try do move stuff with this approach of splitting BC into remote and local and then go for this large renaming :)

trflynn89 commented on Userland/Services/WebDriver/Client.cpp:
```diff
@@ -335,6 +335,14 @@ Web::WebDriver::Response Client::get_window_handles(Web::WebDriver::Parameters p
+Web::WebDriver::Response Client::new_window(Web::WebDriver::Parameters parameters, JsonValue)
+{
+    auto session = TRY(find_session_with_id(parameters[0]));
+    return session->web_content_connection().new_window();
```
Even though the remote end isn't currently parsing it, let's send the payload over anyways. Will make the patch needed to parse them later smaller.

kalenikaliaksandr replied:
done

trflynn89 commented on Userland/Services/WebContent/WebDriverClient.ipc:
```diff
@@ -17,6 +17,7 @@ endpoint WebDriverClient {
     get_title() => (Web::WebDriver::Response response)
     new_window() => (Web::WebDriver::Response response)
     close_window() => (Web::WebDriver::Response response)
+    get_window_handle() => (String handle)
```
Nit: To keep everything in spec order, this belongs above `close_window`

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -365,6 +365,11 @@ void WebDriverConnection::set_is_webdriver_active(bool is_webdriver_active)
     m_page_client.page().set_is_webdriver_active(is_webdriver_active);
 }
 
+Messages::WebDriverClient::GetWindowHandleResponse WebDriverConnection::get_window_handle()
```
Same re: spec order

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.h:
```diff
@@ -51,6 +51,7 @@ class WebDriverConnection final
     virtual Messages::WebDriverClient::BackResponse back() override;
     virtual Messages::WebDriverClient::ForwardResponse forward() override;
     virtual Messages::WebDriverClient::RefreshResponse refresh() override;
+    virtual Messages::WebDriverClient::GetWindowHandleResponse get_window_handle() override;
```
Same re: spec order

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -522,6 +522,45 @@ Messages::WebDriverClient::GetTitleResponse WebDriverConnection::get_title()
     return title;
 }
 
+// 11.5 New Window, https://w3c.github.io/webdriver/#dfn-new-window
```
Nit: To keep everything in spec order, this belongs below close_window

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.h:
```diff
@@ -52,6 +52,7 @@ class WebDriverConnection final
     virtual Messages::WebDriverClient::ForwardResponse forward() override;
     virtual Messages::WebDriverClient::RefreshResponse refresh() override;
     virtual Messages::WebDriverClient::GetTitleResponse get_title() override;
+    virtual Messages::WebDriverClient::NewWindowResponse new_window() override;
```
Same re: spec order

trflynn89 commented on Userland/Services/WebContent/WebDriverClient.ipc:
```diff
@@ -15,6 +15,7 @@ endpoint WebDriverClient {
     forward() => (Web::WebDriver::Response response)
     refresh() => (Web::WebDriver::Response response)
     get_title() => (Web::WebDriver::Response response)
+    new_window() => (Web::WebDriver::Response response)
```
Same re: spec order

trflynn89 commented on Userland/Libraries/LibWeb/WebDriver/Client.h:
```diff
@@ -49,6 +49,7 @@ class Client : public Core::Object {
 
     // 11. Contexts, https://w3c.github.io/webdriver/#contexts
     virtual Response get_window_handle(Parameters parameters, JsonValue payload) = 0;
+    virtual Response new_window(Parameters parameters, JsonValue payload) = 0;
```
Same re: spec order

trflynn89 commented on Userland/Libraries/LibWeb/WebDriver/Client.cpp:
```diff
@@ -65,6 +65,7 @@ static constexpr auto s_webdriver_endpoints = Array {
     ROUTE(DELETE, "/session/:session_id/window"sv, close_window),
     ROUTE(POST, "/session/:session_id/window"sv, switch_to_window),
     ROUTE(GET, "/session/:session_id/window/handles"sv, get_window_handles),
+    ROUTE(POST, "/session/:session_id/window/new"sv, new_window),
```
Same re: spec order

trflynn89 commented on Userland/Services/WebDriver/Client.h:
```diff
@@ -49,6 +49,7 @@ class Client final : public Web::WebDriver::Client {
     virtual Web::WebDriver::Response refresh(Web::WebDriver::Parameters parameters, JsonValue payload) override;
     virtual Web::WebDriver::Response get_title(Web::WebDriver::Parameters parameters, JsonValue payload) override;
     virtual Web::WebDriver::Response get_window_handle(Web::WebDriver::Parameters parameters, JsonValue payload) override;
+    virtual Web::WebDriver::Response new_window(Web::WebDriver::Parameters parameters, JsonValue payload) override;
```
Same re: spec order (though on this side of the IPC connection, belongs below get_window_handles)

kalenikaliaksandr commented:
> I was using Serenity's Browser (not Ladybird) if that makes a difference.

The problem is that initially I implemented handling of IPC call to open a new tab only for Ladybird. Pushed another commit with support for the Browser.

Also when I was testing locally I had to hardcode `m_should_block_pop_ups` to false so `choose_a_browsing_context` can proceed with opening a new top-level BC. I added new commit to allow pop-ups when WebDriver is connected.

Thank you so much for making review and testing!

trflynn89 commented:
Hmm so now I'm getting issues on Browser where we hang when the `on_new_tab` callback tries to get the handle. I issue `driver.post('/session/{session_id}/window/new')`, then we get to the `view().on_new_tab` callback in Browser/Tab.cpp, then nothing progresses.

Maybe we could create the handle ahead of time, and send it to the new tab asynchronously. So in `BrowsingContext::choose_a_browsing_context`, something like:
```c++
auto handle = Crypto::generate_random_uuid().release_value_but_fixme_should_propagate_errors();
m_page->client().page_did_request_new_tab(handle);

chosen = RemoteBrowsingContext::create_a_new_remote_browsing_context(handle);
```

Then in Tab.cpp, after we create a new tab, we send the handle over IPC:
```c++
view().on_new_tab = [this](auto handle) {
    auto& tab = this->window().create_new_tab(URL("about:blank"), true);
    tab.view().set_webdriver_handle(tab.view().handle());
};
```

Not sure what else to do :thinking:  this will interfere with the way the UUID is currently set in `BrowsingContext::create_a_new_browsing_context`, so some rearrangement may need to happen.

trflynn89 commented on Ladybird/WebContentView.cpp:
```diff
@@ -936,7 +936,7 @@ void WebContentView::notify_server_did_close_browsing_context(Badge<WebContentCl
 String WebContentView::notify_request_open_new_tab(Badge<WebContentClient>)
 {
-    return {};
+    return on_new_tab();
```
We typically protect callback invocations with an `if` (see the callbacks below this one):
```c++
if (on_new_tab)
    return on_new_tab();
return {};
```

trflynn89 commented on Userland/Libraries/LibWebView/OutOfProcessWebView.cpp:
```diff
@@ -58,6 +58,11 @@ void OutOfProcessWebView::handle_web_content_process_crash()
 
+String OutOfProcessWebView::notify_request_open_new_tab(Badge<WebContentClient>)
+{
+    return on_new_tab();
```
Same here re: protecting the invocation with an `if`

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -536,6 +536,45 @@ Messages::WebDriverClient::CloseWindowResponse WebDriverConnection::close_window
+    result.set("type"sv, JsonValue { type });
+
+    // 9. Return success with data result.
+    return JsonValue { browsing_context->webdriver_handle() };
```
This is supposed to be `return result;`

trflynn89 commented on Userland/Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -324,6 +324,9 @@ ErrorOr<NonnullRefPtr<WebDriverConnection>> WebDriverConnection::connect(Web::Pa
     auto socket = TRY(Core::LocalSocket::connect(webdriver_ipc_path));
 
+    // Allow pop-ups, or otherwise /window/new won't be able to open a new tab.
+    page_client.page().set_should_block_pop_ups(false);
```
Can you explain this more in the commit message? Why does blocking popups prevent new tabs from opening?

kalenikaliaksandr replied:
sure, done

kalenikaliaksandr commented:
@trflynn89 yes, that is what I meant here https://github.com/SerenityOS/serenity/pull/17855#discussion_r1137155100
I solved that initially in Ladybird by queries handle before any other IPC calls but not sure if I can do it the same way in Browser. I will check! thank you for having a look again!

(for some reason it worked for me in Ladybird yesterday after I applied https://github.com/SerenityOS/serenity/pull/17855#discussion_r1136886866. that is weird. I expected it to start hanging.)

trflynn89 commented:
Ladybird and Browser IPC are *very* different. They both use sockets, but the event loop in Ladybird is driven by Qt is kinda weird ways, so I'm not surprised they behave differently :)

trflynn89 commented:
@kalenikaliaksandr ok so I think what we should do is have `ViewImplementation::create_client` create the window handle, asynchronously send it over IPC, and also store that handle locally in its `m_client_state`. This way, we inform new tabs of their window handles the same way for both the initial tab and WebDriver-created tabs. And when the `on_new_tab` callback is invoked, we already have the new window handle to send back to the `WebDriverConnection`.

Here's a patch that worked for me for Browser (it doesn't do the same update to Ladybird):
https://gist.github.com/trflynn89/fdd1bda758bf07427574fbefdbf94b28

trflynn89 commented:
Ah shoot that doesn't quick work for Ladybird due do a different order of init. On Browser, the order is:
1. Start new WebContent process
2. Send the window handle to the WebContent process
3. Tell WebContent to connect to WebDriver

On Ladybird, it's:
1. Start new WebContent process
2. Tell WebContent to connect to WebDriver
3. Send the window handle to the WebContent process

Because Ladybird connects to WebDriver in it's WebContent's main.cpp. I'll open a PR that consolidates these so they're the same order, than the approach in my previous comment should work fine

trflynn89 commented:
Ok once that other PR is merged, this works now. Sorry for all the comments in a row :smile: 

Here's the updated patch that works for Browser and Ladybird:
https://gist.github.com/trflynn89/fdd1bda758bf07427574fbefdbf94b28
